presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java

```diff
@@ -0,0 +1,94 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.execution.buffer;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.block.BlockEncodingSerde;
+import io.airlift.compress.Compressor;
+import io.airlift.compress.Decompressor;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+
+import java.util.Optional;
+
+import static com.facebook.presto.execution.buffer.PageCompression.UNCOMPRESSED;
+import static com.facebook.presto.execution.buffer.PagesSerdeUtil.readRawPage;
+import static com.facebook.presto.execution.buffer.PagesSerdeUtil.writeRawPage;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.compress.lz4.Lz4RawCompressor.maxCompressedLength;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public class PagesSerde
+{
+    private static final double MINIMUM_COMPRESSION_RATIO = 0.8;
+
+    private final BlockEncodingSerde blockEncodingSerde;
+    private final Optional<Compressor> compressor;
+    private final Optional<Decompressor> decompressor;
+    private final PageCompression compression;
+
+    public PagesSerde(BlockEncodingSerde blockEncodingSerde, PageCompression compression, Optional<Compressor> compressor, Optional<Decompressor> decompressor)
+    {
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.compressor = requireNonNull(compressor, "compressor is null");
+        this.decompressor = requireNonNull(decompressor, "decompressor is null");
+        checkArgument(compressor.isPresent() == decompressor.isPresent(), "compressor and decompressor must both be present or both be absent");
+        this.compression = requireNonNull(compression, "compression is null");
+        checkArgument(compressor.isPresent() || compression == UNCOMPRESSED, "compressor and decompressor must be set if codec is configured");
+    }
+
+    public SerializedPage serialize(Page page)
+    {
+        SliceOutput serializationBuffer = new DynamicSliceOutput(toIntExact((page.getSizeInBytes() + 4)));
+        writeRawPage(page, serializationBuffer, blockEncodingSerde);
+
+        if (!compressor.isPresent()) {
+            return new SerializedPage(serializationBuffer.slice(), UNCOMPRESSED, page.getPositionCount(), serializationBuffer.size());
+        }
+
+        int maxCompressedLength = maxCompressedLength(serializationBuffer.size());
+        byte[] compressionBuffer = new byte[maxCompressedLength];
```

Comment:
Since this object is single threaded and reused for the whole operator (and you do a copy anyway at the end), I would reuse this buffer.

Also use the trick where the field is initialized with `new byte[0]`